src/credentials/GroupDataProviderImpl.h

```diff
@@ -44,123 +44,164 @@ class GroupDataProviderImpl : public GroupDataProvider
+        EndpointIteratorImpl(GroupDataProviderImpl & provider, FabricIndex fabric_index);
         size_t Count() override;
         bool Next(GroupEndpoint & output) override;
         void Release() override;
 
     private:
         GroupDataProviderImpl & mProvider;
-        chip::FabricIndex mFabric = kUndefinedFabricIndex;
-        chip::GroupId mFirstGroup = kUndefinedGroupId;
-        uint16_t mGroup           = 0;
-        size_t mGroupIndex        = 0;
-        size_t mGroupCount        = 0;
-        uint16_t mEndpoint        = 0;
-        size_t mEndpointIndex     = 0;
-        size_t mEndpointCount     = 0;
+        FabricIndex mFabric   = kUndefinedFabricIndex;
+        GroupId mFirstGroup   = kUndefinedGroupId;
+        uint16_t mGroup       = 0;
+        size_t mGroupIndex    = 0;
+        size_t mGroupCount    = 0;
+        uint16_t mEndpoint    = 0;
+        size_t mEndpointIndex = 0;
+        size_t mEndpointCount = 0;
+        bool mFirstEndpoint   = true;
+    };
+
+    class GroupKeyContext : public Crypto::SymmetricKeyContext
+    {
+    public:
+        GroupKeyContext() = default;
+        CHIP_ERROR SetKey(const ByteSpan & value);
+        void Clear();
+
+        CHIP_ERROR EncryptMessage(MutableByteSpan & plaintext, const ByteSpan & aad, const ByteSpan & nonce,
+                                  MutableByteSpan & out_mic) const override;
+        CHIP_ERROR DecryptMessage(MutableByteSpan & ciphertext, const ByteSpan & aad, const ByteSpan & nonce,
+                                  const ByteSpan & mic) const override;
+        CHIP_ERROR EncryptPrivacy(MutableByteSpan & plaintext, const ByteSpan & nonce) const override;
+        CHIP_ERROR DecryptPrivacy(MutableByteSpan & ciphertext, const ByteSpan & nonce) const override;
+
+    protected:
+        uint8_t mKey[Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
     };
 
     class KeySetIteratorImpl : public KeySetIterator
     {
     public:
-        KeySetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
+        KeySetIteratorImpl(GroupDataProviderImpl & provider, FabricIndex fabric_index);
         size_t Count() override;
         bool Next(KeySet & output) override;
         void Release() override;
 
     private:
         GroupDataProviderImpl & mProvider;
-        chip::FabricIndex mFabric = kUndefinedFabricIndex;
-        uint16_t mNextId          = 0;
-        size_t mCount             = 0;
-        size_t mTotal             = 0;
+        FabricIndex mFabric = kUndefinedFabricIndex;
+        uint16_t mNextId    = 0;
+        size_t mCount       = 0;
+        size_t mTotal       = 0;
+    };
+
+    class GroupSessionIteratorImpl : public GroupSessionIterator
+    {
+    public:
+        GroupSessionIteratorImpl(GroupDataProviderImpl & provider, uint16_t session_id);
+        size_t Count() override;
+        bool Next(GroupSession & output) override;
+        void Release() override;
+
+    private:
```

Comment:
Suggest `protected` for all of these, to allow overriding classes to make use of the state.

Reply:
Using protected now.